downgrade

```diff
@@ -105,33 +103,63 @@ prompt_to_ignore() {
   done
 }
 
+filter_packages() {
+  local -n p=$1
+  local name="$2" operator="$3" version="$4"
+  local version_regex="[^-]+-[0-9.]+"
+  # loop over all paths and check manually
+  for path in "${p[@]}"; do
+    path_version=$(sed -r "s/.*$name-($version_regex)-(any|$DOWNGRADE_ARCH)\\.pkg\\.tar\\.(gz|xz|zst)/\1/g" <<<"$path")
+    result=$(vercmp "$path_version" "$version")
+    if [[ "$operator" = ">=" ]] && [[ "$result" -ge "0" ]]; then
+      echo "$path"
+    elif [[ "$operator" = ">" ]] && [[ "$result" -eq "1" ]]; then
+      echo "$path"
+    elif [[ "$operator" = "<=" ]] && [[ "$result" -le "0" ]]; then
+      echo "$path"
+    elif [[ "$operator" = "<" ]] && [[ "$result" -eq "-1" ]]; then
+      echo "$path"
+    fi
+  done
+}
+
 search_packages() {
-  local name version pkgfile_re index
+  local name="$1" operator="$2" version="$3" non_regex_version pkgfile_re index
+  local all_paths_ala all_paths_cached
 
-  if [[ "$1" =~ -([0-9R].*)$ ]]; then
-    version="${BASH_REMATCH[1]}"
-    name=${1%-$version}
+  if [[ "$operator" = "=" ]]; then
     version="$version.*"
   else
+    non_regex_version="$version"
     version="[^-]+-[0-9.]+"
-    name=$1
   fi
   pkgfile_re="$name-$version-(any|$DOWNGRADE_ARCH)\\.pkg\\.tar\\.(gz|xz|zst)"
```

Comment:
I would offer three broad suggestions, which will (unfortunately) impact a lot of different parts of this diff, so I'll comment it here and submit this review and see what you think before reviewing any more:

**1- Commit to owning filtering the versions yourself, and do so in just one place**

That means, for example, this regex-based `find` now becomes simply

```sh
find $PACMAN_CACHE -maxdepth 1 -name "$name-*.pkg.*"
```

(I don't care about being so specific with the extension.)

**2- Don't break `name-operator-version` as early as you are**

It seems to just require passing 3 variables around and doing repeated conditionals. Instead, I would keep the argument as a single thing as much as possible, and pass it around until it gets to the "one place" I mention in (1). Then (and only then) split it and deal with it accordingly. This will require some version of (3) to work well, I think.

(OK, maybe you need to break out `name` to do the `find` I mention in (1), but I would do that right there where you need it, instead of accepting it pre-broken into this area of the script, personally).

**3- Write `filter_packages` as a, well, filter**

```sh
filter_packages()
  local name_version_filter=$1 pkg

  while read -r pkg; do
    if matches_name_version_filter "$name_version_filter" "$pkg"; then
      echo "$pkg"
    fi
  done
}

matches_name_version_filter() {
  local name_version_filter=$1 pkg=$2

  case name_version_filter in
    *>=*)
      # ...
      ;;
    *<=*)
      # ...
      ;;
    *=*)
      # ...
      ;;
    *)
      # assume name[-version]
    ;;
  esac
}
```

Now, slotting that in here becomes a far nicer read (and smaller diff):

```sh
find $PACMAN_CACHE -maxdepth 1 -name "$name-*.pgk.tar.(gz|xz|zst)" |
  filter_packages "$name_version_filter"
```

Reply:
> **1- Commit to owning filtering the versions yourself, and do so in just one place**
> 
> That means, for example, this regex-based `find` now becomes simply
> 
> ```shell
> find $PACMAN_CACHE -maxdepth 1 -name "$name-*.pkg.*"
> ```
>
> (I don't care about being so specific with the extension.)

As far as I understand (and also from experimentation), the currently used explicit regular expression is required in order to prevent fuzzy matches. For example ` $name-*.pkg.*` might result in `vim` matching with `vim-runtime`. I kept it explicit to avoid this.

> **2- Don't break `name-operator-version` as early as you are**
> 
> It seems to just require passing 3 variables around and doing repeated conditionals. Instead, I would keep the argument as a single thing as much as possible, and pass it around until it gets to the "one place" I mention in (1). Then (and only then) split it and deal with it accordingly. This will require some version of (3) to work well, I think.
> 
> (OK, maybe you need to break out `name` to do the `find` I mention in (1), but I would do that right there where you need it, instead of accepting it pre-broken into this area of the script, personally).

Some degree of splitting is already necessary early on in `main` and `process_term` in order to send the clean package names (without operator/version) to `stderr` in case of an error and to send clean package names into `to_ignore`, `previously_installed` and `currently_installed`. I managed to do this less intrusively by working with a single array called `name_op_ver`, which keeps function arguments to single ones.

> 
> **3- Write `filter_packages` as a, well, filter**
> 
> ```shell
> filter_packages()
>   local name_version_filter=$1 pkg
> 
>   while read -r pkg; do
>     if matches_name_version_filter "$name_version_filter" "$pkg"; then
>       echo "$pkg"
>     fi
>   done
> }
> 
> matches_name_version_filter() {
>   local name_version_filter=$1 pkg=$2
> 
>   case name_version_filter in
>     *>=*)
>       # ...
>       ;;
>     *<=*)
>       # ...
>       ;;
>     *=*)
>       # ...
>       ;;
>     *)
>       # assume name[-version]
>     ;;
>   esac
> }
> ```
> 
> Now, slotting that in here becomes a far nicer read (and smaller diff):
> 
> ```shell
> find $PACMAN_CACHE -maxdepth 1 -name "$name-*.pgk.tar.(gz|xz|zst)" |
>   filter_packages "$name_version_filter"
> ```

Thank you for this suggestion. I incorporated this into the code and it looks much cleaner now. Let me know your thoughts.